In [139]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

In [155]:
# import train dataset
train = pd.read_csv('train.csv')
train.head()

,Unnamed: 0,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,1,221900.0,3,1.00,1180,5650,1.0,0,0,3,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,2,538000.0,3,2.25,2570,7242,2.0,0,0,3,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,3,180000.0,2,1.00,770,10000,1.0,0,0,3,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,4,604000.0,4,3.00,1960,5000,1.0,0,0,5,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,5,510000.0,3,2.00,1680,8080,1.0,0,0,3,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503


In [152]:
# import test dataset
test = pd.read_csv('test.csv')
test.head()

,Unnamed: 0,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,1001,3131200640,20150427T000000,700000.0,4,2.00,1830,4590,2.0,0,...,8,1830,0,1908,0,98105,47.6593,-122.327,1650,4590
1,1002,984000710,20141022T000000,270000.0,3,2.00,1560,8853,1.0,0,...,7,1560,0,1967,0,98058,47.4312,-122.171,1610,8750
2,1003,4167300350,20140508T000000,258000.0,4,1.75,1730,8320,1.0,0,...,7,1230,500,1977,0,98023,47.3270,-122.361,1840,9800
3,1004,2826049282,20140614T000000,530000.0,3,2.50,1930,7214,2.0,0,...,8,1930,0,2005,0,98125,47.7191,-122.309,1930,7266
4,1005,8946750030,20141218T000000,245000.0,3,2.25,1422,3677,2.0,0,...,7,1422,0,2012,0,98092,47.3204,-122.178,1677,3677


In [156]:
# TRAIN:
# ignore columns: id, date and zipcode
# also ignore price column because this is the response and what we are trying to predict
columns_to_drop = ['Unnamed: 0', 'zipcode', 'price']
x_train = train.drop(columns=columns_to_drop)
x_train.head()

,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,lat,long,sqft_living15,sqft_lot15
0,3,1.00,1180,5650,1.0,0,0,3,7,1180,0,1955,0,47.5112,-122.257,1340,5650
1,3,2.25,2570,7242,2.0,0,0,3,7,2170,400,1951,1991,47.7210,-122.319,1690,7639
2,2,1.00,770,10000,1.0,0,0,3,6,770,0,1933,0,47.7379,-122.233,2720,8062
3,4,3.00,1960,5000,1.0,0,0,5,7,1050,910,1965,0,47.5208,-122.393,1360,5000
4,3,2.00,1680,8080,1.0,0,0,3,8,1680,0,1987,0,47.6168,-122.045,1800,7503


In [153]:
# TEST:
# ignore columns: id, date and zipcode
# also ignore price column because this is the response and what we are trying to predict
columns_to_drop = ['Unnamed: 0', 'id', 'date', 'zipcode', 'price']
x_test = test.drop(columns=columns_to_drop)
x_test.head()

,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,lat,long,sqft_living15,sqft_lot15
0,4,2.00,1830,4590,2.0,0,0,3,8,1830,0,1908,0,47.6593,-122.327,1650,4590
1,3,2.00,1560,8853,1.0,0,0,3,7,1560,0,1967,0,47.4312,-122.171,1610,8750
2,4,1.75,1730,8320,1.0,0,0,3,7,1230,500,1977,0,47.3270,-122.361,1840,9800
3,3,2.50,1930,7214,2.0,0,0,3,8,1930,0,2005,0,47.7191,-122.309,1930,7266
4,3,2.25,1422,3677,2.0,0,0,3,7,1422,0,2012,0,47.3204,-122.178,1677,3677


In [157]:
# scale the data so that each feature has a mean of 0 and standard deviation of 1
scaler = StandardScaler()

# scale train dataset
for col in x_train:
    train_scaled = x_train[col].values.reshape(-1, 1) #reshape
    x_train[col] = scaler.fit_transform(train_scaled) 

# scale test dataset
for col in x_test:
    test_scaled = x_test[col].values.reshape(-1, 1) #reshape
    x_test[col] = scaler.fit_transform(test_scaled) 


# check mean and standard deviation of each feature to double check that scaling worked
for feature in x_train:
    mean = x_train.mean()
    std = x_train.var()

for feature in x_test:
    mean = x_test.mean()
    std = x_test.var()

# organize all variables into a table
stats_train = {
    'mean': mean,
    'std': std
}

stats_test = {
    'mean': mean,
    'std': std
}

feature_data_train = pd.DataFrame(stats_train)
feature_data_test = pd.DataFrame(stats_test)
feature_data_test


,mean,std
bedrooms,-1.385558e-16,1.001001
bathrooms,2.122746e-16,1.001001
sqft_living,-2.113865e-16,1.001001
sqft_lot,-1.421085e-17,1.001001
floors,-8.704149e-17,1.001001
waterfront,-2.842171e-17,1.001001
view,-8.792966e-17,1.001001
condition,1.776357e-16,1.001001
grade,-3.215206e-16,1.001001
sqft_above,2.842171e-17,1.001001


In [158]:
# divide the price by 1000 for all rows in the train and test datasets
y_train = train['price'] / 1000
y_test = test['price'] / 1000

In [160]:
# include a column of ones into features for the bias
# lecture slides showed column of ones in features matrix:
shape = x_train.shape[0]
x_train = np.append(x_train, np.ones((shape,1)), axis=1)

# test dataset:
shape_test = x_test.shape[0]
matrix_test = np.append(x_test, np.ones((shape,1)), axis=1)

1. Write code for gradient descent for training linear regression using the algorithm from class.

In [146]:
# gradient descent for training linear regression
# algorithm: thetaj <- thetaj - alpha*deritivative of J(theta) - alpha is learning rate (small)
# algorithm should stop when: 1. update in theta is below some threshold or 2. maximum number of iterations reached (max_iterations var)
# choose initial value for theta - choose new values for theta to reduce J(theta) -> theta should get closer to minimum?

# gradient = slope of line tangent to curve

# features = x, price = y (both matrices?)
def gradient_descent(features, price, alpha, max_iterations):
    
    # initialize theta to 0?
    # length of theta vector should be amt of features + 1?
    theta = np.zeros(features.shape[1])
    N = features.shape[0]

    for iteration in range(max_iterations):
        # 1. get predicted values
        y_pred = features @ theta

        # 2. calc error difference btwn predicted and actual?
        error = y_pred - price

        # 3. gradient: 1/N * X.T * (error calculated above) 
        gradient = (1/N) * (features.T @ error)

        # 4. update:
        theta = theta - (alpha * gradient)

    return theta

In [164]:
train = gradient_descent(x_train, y_train, 0.01, 10)
test = gradient_descent(x_test, y_test, 0.01, 10)

In [167]:
def predict_response(x, theta):
    return x @ theta

pred = predict_response(x_train, train)
pred2 = predict_response(x_test, test)
r2_score(y_test, pred2)

-1.4256760191017714

2. Vary the value of the learning rate (at least 3 different values $\alpha \in \{0.01,0.1,0.5\}$) and report the value of the model parameter $\theta$ after different number of iterations (10, 50, and 100). Include in a table the MSE and $R^2$ metrics on the training and testing set for the different number of iterations and different learning rates. You can choose more values of the learning rates to observe how the
behavior of the algorithm changes.

In [149]:
alphas = [0.01, 0.1, 0.5]
max_iterations = [10, 50, 100]
mses = []
r_squared = []

for alpha in alphas:
    mse1 = []
    r2_1 = []
    for iteration in max_iterations:

        grad_desc = gradient_descent(x_train, y_train, alpha, iteration)

        y_pred = predict_response(x_train, grad_desc)

        mse1.append(mean_squared_error(y_train, y_pred))
        r2_1.append(r2_score(y_train, y_pred))
    
    mses.append(mse1)
    r_squared.append(r2_1)

In [151]:
# organize all variables into a table
# TRAIN:

data = []

for alpha in range(0, len(alphas)):
    
    stats_train = { 
        'alpha': alphas[alpha],
        'max iterations': max_iterations,
        'MSE metric': mses[alpha],
        'R squared metric': r_squared[alpha]
    }

    data.append(pd.DataFrame(stats_train))
print("Training data:")
data

Training data:


[   alpha  max iterations     MSE metric  R squared metric
 0   0.01              10  294798.733591         -1.560413
 1   0.01              50  138295.915198         -0.201140
 2   0.01             100   70118.986603          0.390996,
    alpha  max iterations    MSE metric  R squared metric
 0    0.1              10  66499.315474          0.422434
 1    0.1              50  31578.978073          0.725727
 2    0.1             100  31497.692326          0.726433,
    alpha  max iterations    MSE metric  R squared metric
 0    0.5              10  6.118299e+08     -5.312921e+03
 1    0.5              50  1.649496e+25     -1.432635e+20
 2    0.5             100  5.698752e+45     -4.949532e+40]

In [161]:
alphas = [0.01, 0.1, 0.5]
max_iterations = [10, 50, 100]
mses = []
r_squared = []

for alpha in alphas:
    mse1 = []
    r2_1 = []
    for iteration in max_iterations:

        grad_desc = gradient_descent(x_test, y_test, alpha, iteration)

        y_pred = predict_response(x_test, grad_desc)

        mse1.append(mean_squared_error(y_test, y_pred))
        r2_1.append(r2_score(y_test, y_pred))
    
    mses.append(mse1)
    r_squared.append(r2_1)


# organize all variables into a table
# TEST:

data = []

for alpha in range(0, len(alphas)):
    
    stats_test = { 
        'alpha': alphas[alpha],
        'max iterations': max_iterations,
        'MSE metric': mses[alpha],
        'R squared metric': r_squared[alpha]
    }

    data.append(pd.DataFrame(stats_test))
print("Test data:")
data

Test data:


[   alpha  max iterations     MSE metric  R squared metric
 0   0.01              10  404425.560193         -1.425676
 1   0.01              50  359637.751368         -1.157046
 2   0.01             100  352219.079873         -1.112550,
    alpha  max iterations     MSE metric  R squared metric
 0    0.1              10  351869.150086         -1.110452
 1    0.1              50  347857.699096         -1.086392
 2    0.1             100  347623.288455         -1.084986,
    alpha  max iterations    MSE metric  R squared metric
 0    0.5              10  7.499447e+08     -4.497041e+03
 1    0.5              50  1.281918e+25     -7.688725e+19
 2    0.5             100  2.506978e+45     -1.503643e+40]

3. Write some observations about the behavior of the algorithm: How do the metrics change with different learning rates;How many
iterations are needed; Does the algorithm converge to the optimal solution, etc.

Different learning rates show underfitting, converging to the optimal solution and overfitting. At the lowest learning rate of 0.01, the MSE metrics are high and the R squared metrics are low, showing that the model is taking too small steps to learn the data. Conversely, at the highest learning rate of 0.5, the MSE metrics and R squared metris are exteremly high and the R squared metrics are extremely high negative numbers as well. This shows overfitting. Whereas the learning rate of 0.1, shows the most optimal solutions with the lowest MSE metrics and the best R squared values. 

For the training dataset the learning rate of 0.1 and max iterations of 100, converges to the optimal solution with MSE of 31497.692326 and R squared metric of 0.726433. For the testing dataset, it seems like the most optimal solution has the same learning rate and max iterations, as this as the lowest MSE metric.

Overall it seems like there are better results with more iterations, which makes sense as the model needs more time to learn the data.